pcdsdevices/mv_interface.py

```diff
@@ -120,3 +134,352 @@ def umvr(self, delta, timeout=None):
             will be use.
         """
         self.umv(delta + self.wm(), timeout=timeout)
+
+
+def setup_preset_paths(**paths):
+    """
+    Prepare the `Presets` class with the correct paths for saving and loading
+    presets.
+
+    Parameters
+    ----------
+    **paths: ``str`` keyword args
+        A mapping from type of preset to destination path. These will be
+        directories that contain the yaml files that define the preset
+        positions.
+    """
+    Presets._paths = {}
+    for k, v in paths.items():
+        Presets._paths[k] = Path(v)
+
+
+class Presets:
+    """
+    Manager for device preset positions. This provides methods for adding new
+    presets, checking which presets are active, and related utilities.
+
+    It will install the ``mv_presetname`` and ``wm_presetname`` methods onto
+    the associated device.
+
+    Parameters
+    ----------
+    device: ``Device``
+        The device to manage saved preset positions for. It must implement the
+        `FltMvInterface`.
+
+    Attributes
+    ----------
+    positions: ``SimpleNamespace``
+        A namespace that contains all of the active presets as `PresetPosition`
+        objects.
+    """
+    _paths = {}
+
+    def __init__(self, device):
+        self._device = device
+        self._methods = []
+        self.name = device.name + '_presets'
+        self.sync()
+
+    def _path(self, preset_type):
+        """
+        Utility function go get the proper ``Path`` object that points to the
+        presets file.
+        """
+        path = self._paths[preset_type] / (self._device.name + '.yml')
+        logger.debug('select presets path %s', path)
+        return path
+
+    def _read(self, preset_type):
+        """
+        Utility function to get a particular preset's datum dictionary.
+        """
+        logger.debug('read presets for %s', self._device.name)
+        with self._path(preset_type).open('r') as f:
+            return yaml.load(f)
+
+    def _write(self, preset_type, data):
+        """
+        Utility function to overwrite a particular preset's datum dictionary.
+        """
+        logger.debug('write presets for %s', self._device.name)
+        with self._path(preset_type).open('w') as f:
+            yaml.dump(data, f, default_flow_style=False)
```

Comment:
Consider using [flock](https://docs.python.org/3.6/library/fcntl.html#fcntl.flock) to make sure that we don't have two sessions try and write to the file simultaneously?

Reply:
Is it possible for `flock` to clean up poorly and leave the file in a locked state? As implemented it will only cause issues if two sessions in one hutch are writing to the same object's presets at the same time +/- a second or two. I'd be wrapping the `_update` method in some sort of `flock` wrapper so that each client would wait for the flock, then read/update/write.

Reply:
Good point. Might be overkill. You can use it as a context manager and I assume since it is in standard lib it should handle most cases. Your are correct it might be overkill.

Reply:
I'll do this to protect ourselves when we decide to assign presets programatically as you described below

Reply:
Yeah used in the `happi` JSON interface and no issues yet.

https://github.com/pcdshub/happi/blob/a1443cd76140aa562c40f86a4e13b301ffd16755/happi/backends/json_db.py#L91

Reply:
This ended up being very difficult to implement correctly.